File: packages/core/src/Tunnel.ts
import { randomId } from "@tgrok/utils";
import { event } from "./event";
import { TunnelStatus } from "@tgrok/shared";

class Tunnel {
  public readonly id: string;
  public url = "";
  public localHost: string;
  public localPort: number;
  public requestId: string;

  private readonly hostname: string;
  public readonly subdomain: string;
  public readonly protocol: string;
  private readonly remotePort: number;

  private _status = TunnelStatus.DISCONNECTED;

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  constructor(config: any) {
    this.id = config.id;
    this.requestId = randomId("a0", 20);
    this.hostname = config.hostname;
    this.subdomain = config.subdomain;
    this.protocol = config.protocol;
    this.localHost = config.lhost;
    this.localPort = config.lport;
    this.remotePort = config.rport;
  }

  public set status(val: number) {
    this._status = val;
    event.emit("info", {
      evt: "tunnel:status",
      payload: {
        id: this.id,
        status: this._status,
        url: this.url,
      },
    });
  }

  public get status(): number {
    return this._status;
  }

  public request = (): object => {
    return {
      Type: "ReqTunnel",
      Payload: {
        ReqId: this.requestId,
        Protocol: this.protocol,
        Hostname: this.hostname,
        Subdomain: this.subdomain,
        HttpAuth: "",
        RemotePort: this.remotePort,
      },
    };
  };
}

export {
  Tunnel,
};

File: packages/ios/Tgrok/tgrok/Tunnel.swift
//
//  Tunnel.swift
//  Tgrok
//
//  Created by Yueyu Zhao on 2020/9/13.
//  Copyright © 2020 Yueyu Zhao. All rights reserved.
//

import Foundation
import SwiftyJSON

class Tunnel: NSObject {
    
    let id: String
    var url: String?
    let localHost: String
    let localPort: UInt16
    let requestId = UUID().uuidString
    
    let hostname: String
    let subdomain: String
    let proto: String
    let remotePort: UInt16
    
    var _status = 0
    
    var status: Int {
        set (val) {
            _status = val
            NotificationCenter.default.post(name: .tgrok, object: self.statusJson())
        }
        get {
            return _status
        }
    }
    
    func statusJson() -> JSON {
        return JSON([
            "evt": "tunnel:status",
            "payload": [
                "id": self.id,
                "status": self._status,
                "url": self.url ?? ""
            ],
        ])
    }
    
    func request() -> JSON {
        return JSON([
            "Type": "ReqTunnel",
            "Payload": [
                "ReqId": self.requestId,
                "Protocol": self.proto,
                "Hostname": self.hostname,
                "Subdomain": self.subdomain,
                "HttpAuth": "",
                "RemotePort": self.remotePort
            ],
        ])
    }
    
    init(_ config: JSON) {
        self.id = config["id"].stringValue
        self.hostname = config["hostname"].stringValue
        self.subdomain = config["subdomain"].stringValue
        self.proto = config["protocol"].stringValue
        self.localHost = config["localHost"].stringValue
        self.localPort = config["localPort"].uInt16Value
        self.remotePort = 0
        super.init()
    }
    
}
